src/dbnode/network/server/tchannelthrift/compress_client_server.go

```diff
@@ -0,0 +1,273 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannelthrift
+
+import (
+	"sync"
+
+	"github.com/m3db/m3/src/dbnode/generated/thrift/rpc"
+	"github.com/m3db/m3/src/x/instrument"
+
+	apachethrift "github.com/apache/thrift/lib/go/thrift"
+	"github.com/uber-go/tally"
+	"github.com/uber/tchannel-go/thrift"
+)
+
+var _ thrift.TChanClient = (*snappyTChanClient)(nil)
+
+type snappyTChanClient struct {
+	client thrift.TChanClient
+}
+
+// NewSnappyTChanClient creates a new snappy TChanClient.
+func NewSnappyTChanClient(client thrift.TChanClient) thrift.TChanClient {
+	return &snappyTChanClient{client: client}
+}
+
+func (c *snappyTChanClient) Call(
+	ctx thrift.Context,
+	serviceName, methodName string,
+	req, resp apachethrift.TStruct,
+) (bool, error) {
+	snappyReq := snappyTStructPool.Get().(*snappyTStruct)
+	snappyReq.Reset(req, apachethrift.SnappyCompressType)
+
+	snappyResp := snappyTStructPool.Get().(*snappyTStruct)
+	snappyResp.Reset(resp, apachethrift.NoneCompressType)
+
+	result, err := c.client.Call(ctx, serviceName, methodName, snappyReq, snappyResp)
+
+	snappyTStructPool.Put(snappyReq)
+	snappyTStructPool.Put(snappyResp)
+
+	return result, err
+}
+
+var snappyTStructPool = sync.Pool{
+	New: func() interface{} {
+		return &snappyTStruct{}
+	},
+}
+
+var _ apachethrift.TStruct = (*snappyTStruct)(nil)
+
+type snappyTStruct struct {
+	thriftStruct      apachethrift.TStruct
+	writeCompressType apachethrift.CompressType
+}
+
+func (s *snappyTStruct) Reset(
+	thriftStruct apachethrift.TStruct,
+	writeCompressType apachethrift.CompressType,
+) {
+	s.thriftStruct = thriftStruct
+	s.writeCompressType = writeCompressType
+}
+
+func (s *snappyTStruct) Write(p apachethrift.TProtocol) error {
+	compressibleProtocol, ok := p.(apachethrift.TCompressibleProtocol)
+	if ok {
+		opts := apachethrift.WriteCompressibleMessageBeginOptions{
+			CompressType: s.writeCompressType,
+		}
+		err := compressibleProtocol.WriteCompressibleMessageBegin(opts)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := s.thriftStruct.Write(p)
+
+	var writeEndErr error
+	if ok {
+		// Always make sure to read call read message end even if encounter
+		// an error so that cleanup can occur.
```

Comment:
Comment's a bit hard to understand, maybe `Ensure compressible message cleaned up even if error occurred`?